### ISSUE TITLE
"grad can be implicitly created only for scalar outputs" when using grad_cam on my own segmentation model.

### ISSUE THREAD

**Issue by lifuguan** [jacobgil/pytorch-grad-cam#109]
I've been successfully using grad_cam on **the official ResNet code(classifier)**, which works pretty well.
However, when I implement it on **my own segmentation model**, it throws out an error: **grad can be implicitly created only for scalar outputs**. And I have no idea about how to solve it.

The screenshot is shown below:
![image](https://user-images.githubusercontent.com/17793134/124614534-ea31c600-dea6-11eb-9b87-dd16ed6eeccf.png)

Any help will be appreciated!!!!!

**Comment by jacobgil** [comment]
What does model1 return?
Does it return classification scores, the same as torchvision.resnet34?
Or maybe it's modified to return some tuple?

**Comment by lifuguan** [comment]
> What does model1 return?
> Does it return classification scores, the same as torchvision.resnet34?
> Or maybe it's modified to return some tuple?


MODEL1 is U-Net with attention-gate, whose output is a two-dimension tensor, aiming at image segmentation.

`model1.target_layer` is shown below:
![image](https://user-images.githubusercontent.com/17793134/124691553-935fd700-df0e-11eb-87ec-f8f99500716f.png)

**Comment by jacobgil** [comment]
I think the problem is that indeed the output for the category isn't a scalar, it's a 2D image, since you're using a segmentation network, and it can't compute the gradient for it.
You will have to reduce it to a scalar. Some reasonable choices might be to:
1. Take the center pixel.
2. Average the scores for the categories across all pixels, or in a window around the center pixel. 

To do that you will have to modify the code where the loss is computed:
https://github.com/jacobgil/pytorch-grad-cam/blob/master/pytorch_grad_cam/base_cam.py#L37

You can just write another custom get_loss function and replace it with model.get_loss = new_get_loss.
Maybe something like:

```
    def segmentation_get_loss(output, target_category):
        loss = 0
        for i in range(len(target_category)):
            loss = loss + output[i, target_category, :, :].mean()
        return loss
```

**Comment by lifuguan** [comment]
> I think the problem is that indeed the output for the category isn't a scalar, it's a 2D image, since you're using a segmentation network, and it can't compute the gradient for it.
> You will have to reduce it to a scalar. Some reasonable choices might be to:
> 
> 1. Take the center pixel.
> 2. Average the scores for the categories across all pixels, or in a window around the center pixel.
> 
> To do that you will have to modify the code where the loss is computed:
> https://github.com/jacobgil/pytorch-grad-cam/blob/master/pytorch_grad_cam/base_cam.py#L37
> 
> You can just write another custom get_loss function and replace it with model.get_loss = new_get_loss.
> Maybe something like:
> 
> ```
>     def segmentation_get_loss(output, target_category):
>         loss = 0
>         for i in range(len(target_category)):
>             loss = loss + output[i, target_category, :, :].mean()
>         return loss
> ```

Thanks a lot!!!
